### PR TITLE
[Verl] Add entropy loss to `cross_entropy_loss` and `fused_linear_cross_entropy_loss`

### PR DESCRIPTION
**Pull request by hongpeng-guo**
## Summary

In RLHF workflows, such as [verl]([[url](https://github.com/volcengine/verl)]), the actor forward function usually generates both losses of `cross_entropy_loss (-log_probs)` and `entropy_loss`, the later was used to encourage the policy to be not over-deterministic.

There is a real needs for a kernel that will generates both the two losses, without materializing the huge `logits` tensor. Liger-kernel's `fused_linear_cross_entropy_loss` already works well to generate the `cross_entropy_loss`, but only calculating the second part of the loss, i.e., the entropy loss.

This PR adds the `entropy` loss option to the existing FLCE loss, and work as one important step to support verl. 

1. Adding the entropy calculation in the second pass of online softmax in `cross_entropy.py::liger_cross_entropy_kernel`, both the loss and its gradient subject to input are calculated and stored;
2. Propagate the changes to relevant modules in `fused_linear_cross_entropy.py`,
3. Propagate relavent changes to other functional modules in PyTorch interface. 

## Testing Done

Made existing unit tests working; Adding new unittest WIP.

- Hardware Type: <BLANK>
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by Tcc0403** [comment]
Please add a unit test with `return_entropy_loss`. You can write a new pytorch implementation like `CrossEntropyWithZLoss`, or `return_entropy_loss` functionality on top of it.

**Comment by hongpeng-guo** [comment]
Update: Met some numerical unstable issue, inverstigating

**Review by qingquansong — COMMENTED** [review]
Thanks for the efforts! Let's try to test thoroughly on both accuracy (numerical stability) and speed (including old ones) before checking in. Considering we may fused more and more losses such as the existing Z loss and the added entropy loss, api outputs kind of diverged and also make the loss quite heavy with multiple branches coupling together (like label smoothing, target weights, etc) We probably need to refactor a bit to make it cleaner to dev later. cc  @ByronHsu  @shivam15s @Tcc0403 @shimizust

File: src/liger_kernel/ops/cross_entropy.py [hunk 1]
```diff
@@ -167,9 +196,22 @@ def liger_cross_entropy_kernel(
...
+                dX_entropy_ptr + X_offsets,
+                mask=X_offsets < n_cols,
+                other=0.0,
+            )
+        # valid mask for the entropy loss
+        valid_mask = X_offsets < n_cols
+
         if not HAS_WEIGHT:
             # softmax(x_i)
             X_block = tl.exp(X_block - m) / d
```
[the remark this answers is not in the document]
Comment (hongpeng-guo) [reply src/liger_kernel/ops/cross_entropy.py:211]: Sounds good!

File: src/liger_kernel/ops/cross_entropy.py [hunk 2]
```diff
@@ -51,6 +54,7 @@ def liger_cross_entropy_kernel(
 
     Parameters:
     X_ptr: Pointer to input tensor.
+    dX_entropy_ptr: Pointer to tensor to store the gradient of the input w.r.s. to the entropy loss
```
[the remark this answers is not in the document]
Comment (hongpeng-guo) [reply src/liger_kernel/ops/cross_entropy.py:57]: Thanks for catching this!

File: src/liger_kernel/ops/cross_entropy.py [hunk 3]
```diff
@@ -167,9 +196,22 @@ def liger_cross_entropy_kernel(
...
+            )
+        # valid mask for the entropy loss
+        valid_mask = X_offsets < n_cols
+
         if not HAS_WEIGHT:
             # softmax(x_i)
             X_block = tl.exp(X_block - m) / d
+            if RETURN_ENTROPY_LOSS:
+                # derivatives of the entropy loss term
+                dX_entropy_block = tl.where(valid_mask, X_block * (-tl.log(X_block) - entropy_loss), 0.0)
```
[the remark this answers is not in the document]
Comment (hongpeng-guo) [reply src/liger_kernel/ops/cross_entropy.py]: removed the `tl.where` here. I don't think they are necessary, either.

File: src/liger_kernel/ops/cross_entropy.py [hunk 4]
```diff
@@ -248,11 +299,16 @@ def liger_cross_entropy_kernel(
             loss = loss / n_non_ignore
         # TODO: Implement weighted z_loss. Currently, z_loss is not scaled by weight.
         z_loss = z_loss / n_non_ignore
+        # TODO: Implement weighted entropy loss. Currently, entropy loss is not scaled by weight.
+        entropy_loss = entropy_loss / n_non_ignore
```
[the remark this answers is not in the document]
Comment (hongpeng-guo) [reply src/liger_kernel/ops/cross_entropy.py]: Thanks for catching this. I think this is a bug in my program. I just fixed it. But it seems the numerical problem is still there. Maybe we need to take a deeper look.

[the remark this answers is not in the document]
Comment (hongpeng-guo) [reply src/liger_kernel/ops/cross_entropy.py]: BTW, it seems the CI stops running for this PR for some reason.

**Comment by hongpeng-guo** [comment]
close this PR for now.
